### PR TITLE
Vector gate mlp

### PR DESCRIPTION
**Pull request by oli-clive-griffin**
Removes the existing `Gate` class (scalar affine gate), and adds a new `VectorGateMLP` which is a gate which reads from the vector-valued input to the layer being gated. For example, in in MLP up-proj layer, the input to the gate would be the residual stream.

Also:
- bumps torch to 2.6, due to [this](https://nvd.nist.gov/vuln/detail/CVE-2025-32434) vulnerability breaking huggingface `from_pretrained`. 
- fixes a bunch of type errors due to having to bump torch

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -4,6 +4,14 @@
     // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
     "version": "0.2.0",
     "configurations": [
+        {
+            "name": "Python Debugger: Current File with Arguments",
+            "type": "debugpy",
+            "request": "launch",
+            "program": "${file}",
+            "console": "integratedTerminal",
+            "args": "${command:pickArgs}"
```
Comment (danbraunai) [inline .vscode/launch.json:13]: oh my lordy I had no idea this existed

File: spd/configs.py [hunk 2]
```diff
@@ -109,10 +110,13 @@ class Config(BaseModel):
         ...,
         description="Number of stochastic masks to sample when using stochastic recon losses",
     )
-    n_ci_mlp_neurons: NonNegativeInt = Field(
-        default=0,
-        description="Number of hidden neurons in the MLP used to calculate the causal importance."
-        "If 0, use a single-layer gate.",
+    gate_type: GateType = Field(
```
Comment (danbraunai) [inline spd/configs.py:113]: Lee has brought up the good point that we should probably avoid the word "gate" everywhere. We're not actually learning the gates here, we're learning the ci function. The gates, if it's at all a reasonable word, would be the (stochastic) masks.

I know the word gate is very embedded throughout this codebase, but for now I'd like to avoid adding new instances of it, and would probably like to have a PR at some point which removes it everywhere.
I added #48 which asks to remove it throughout the codebase.

File: spd/models/components.py [hunk 3]
```diff
@@ -8,58 +8,67 @@
 
 from spd.utils.module_utils import init_param_
 
+GateType = Literal["mlp", "vector_mlp"]
 
-class Gate(nn.Module):
-    """A gate that maps a single input to a single output."""
 
-    def __init__(self, C: int):
+class ParallelLinear(nn.Module):
```
Comment (danbraunai) [inline spd/models/components.py:14]: This class is always called with a single output dimension in SPD, and I'm not sure if there are use cases where we might want a value other than 1.  I thus think it probably doesn't make sense to have it:
1. It's a little less clear to the reader when they read this class that Do is always 1.
2. I'm worried that operations become less efficient with the extra singleton dimension here. But I don't know, and it might be minimal.

File: spd/models/components.py [hunk 4]
```diff
@@ -8,58 +8,67 @@
...
-        fan_val = 1  # Since each weight gets applied independently
-        init_param_(self.weight, fan_val=fan_val, nonlinearity="linear")
+        self.W_CDiDo = nn.Parameter(torch.empty(C, input_dim, output_dim))
+        self.bias_Do = nn.Parameter(torch.zeros(C, output_dim))
+        init_param_(self.W_CDiDo, fan_val=input_dim, nonlinearity="relu")
 
     @override
-    def forward(self, x: Float[Tensor, "... C"]) -> Float[Tensor, "... C"]:
-        return x * self.weight + self.bias
+    def forward(self, x_BxCDi: Tensor) -> Tensor:
```
Comment (danbraunai) [inline spd/models/components.py:24]: Do you have a plan for switching to suffix types rather than jaxtypes? I think I'm onboard with suffix types after our discussions and more thought, but don't really like that this PR just mixes them both. To save time, I think it's OK to merge this in provided there is another PR that gets merged in the coming days that uses suffixes everywhere.

**Review by danbraunai — CHANGES_REQUESTED** [review]
I'd like to see an evals run on all of tms and resid_mlp with a PR as big as this. It should get similar but not necessarily the same results. I've set one off [here](https://wandb.ai/goodfire/spd/reports/SPD-Run-Report---run_20250715_103042--VmlldzoxMzU5Njc3NQ==).

Apart from that, and a couple of comments, looks great.

File: spd/configs.py (hunk 2 above)
Comment (oli-clive-griffin) [reply spd/configs.py:113]: I think this will make #39 tricky to merge if I change this now. Would you mind if I handled this after merging that?

File: spd/models/components.py (hunk 3 above)
Comment (oli-clive-griffin) [reply spd/models/components.py:14]: `Do` isn't always singleton: with multiple layers the first n-1 have scalar outputs. We Could have seperate implementations for the input (1 -> d in the GateMLP case), intermediate (d -> d), and output (d -> 1) layers but that seems a little overkill.

but yea I can test the performance to make sure this isn't slowing us down

File: spd/models/components.py (hunk 4 above)
Comment (oli-clive-griffin) [reply spd/models/components.py]: yea good point, I'll just keep to jax for now

File: .vscode/launch.json (hunk 1 above)
Comment (oli-clive-griffin) [reply .vscode/launch.json:13]: ikr it's so great

File: spd/configs.py (hunk 2 above)
Comment (danbraunai) [reply spd/configs.py:113]: Yep fair enough, someone can handle after (not on you).

**Comment by oli-clive-griffin** [comment]
huh, weird, by reports I meant those from the evals runs you set off, agree they don't look great, will dig in and try and see whats going on

**Comment by oli-clive-griffin** [comment]
found the problem, it was:
- the final layer wasn't init'd with a nonlinearity of "linear"
- the final layer had a gelu after it

**Comment by oli-clive-griffin** [comment]
@danbraunai looking much better as far as I can tell

**Review by danbraunai — APPROVED** [review]
lgtm
